### DISCUSSION TITLE
AttributeError: 'NoneType' object has no attribute 'get_mask'

### DISCUSSION THREAD

**Discussion by meniua** [raphaelvallat/yasa#104] [Q&A] [answered]
When trying to use YASA to perform single-channel sleep spindles detection(with the sample data of yasa), after'sp = yasa.spindles_detect(data, sf)', it return nonetype:
E:\Software\Anaconda\python.exe E:\python_work\SpindlerDetection\yuan.py 
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s finished
**<class 'NoneType'>**
23-Nov-22 19:43:26 | WARNING | No spindle were found in channel CHAN000.
23-Nov-22 19:43:26 | WARNING | No spindles were found in data. Returning None.
Traceback (most recent call last):
  File "E:\python_work\SpindlerDetection\yuan.py", line 32, in <module>
    mask = sp.get_mask()
**AttributeError: 'NoneType' object has no attribute 'get_mask'**

import yasa
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set(font_scale=1.2)
data = np.loadtxt('E:/yasa-master/notebooks/data_N2_spindles_15sec_200Hz.txt')
sf = 200
times = np.arange(data.size) / sf
sp = yasa.spindles_detect(data, sf)
print(type(sp))
mask = sp.get_mask()
spindles_highlight = data * mask
spindles_highlight[spindles_highlight == 0] = np.nan
fig, ax = plt.subplots(1, 1, figsize=(14, 4))
plt.plot(times, data, lw=1.5, color='tab:grey')
plt.plot(times, spindles_highlight, lw=1.5, color='tab:blue')
plt.xlabel('Time (seconds)')
plt.ylabel('Amplitude (uV)')
plt.xlim([0, times[-1]])
plt.xticks(np.arange(0, 19, 2))
plt.title('Spindles detection')
sns.despine()

**Comment by remrama** [comment]
So weird this happened again. And again when I run this I have no problems, spindles are detected.

I ran the following code on Windows with `yasa 0.6.2`, `mne 1.2.2`, `numpy 1.23.5`, and Python 3.10.6.
```python
import numpy as np
import yasa
sf = 200
data = np.loadtxt('./data_N2_spindles_15sec_200Hz.txt')
sp = yasa.spindles_detect(data, sf)
```

@meniua are you using the same versions? If not could you try that?

**Reply by meniua** [reply]
I ran the code on Windows with yasa 0.6.2, mne 1.2.2, numpy 1.21.5,and Python 3.9.15(The highest version allowed by the Anaconda environment)，is the numpy 1.23.5, and Python 3.10.6 necessary?

**Reply by remrama** [reply]
Not sure... Changing those would definitely be worth trying, I'm not sure where else to look right now. In the meantime I'd say keep an eye on a similar thread #102 where the same problem is being troubleshooted.

**Comment by apavlo89** [comment]
Same issue here. This is my pip list

Package                       Version
----------------------------- ---------
alabaster                     0.7.12
ansi2html                     1.6.0
antropy                       0.1.4
appdirs                       1.4.4
argon2-cffi                   21.3.0
argon2-cffi-bindings          21.2.0
arrow                         1.2.2
astroid                       2.6.6
asttokens                     2.0.5
atomicwrites                  1.4.0
attrs                         21.4.0
autopep8                      1.6.0
Babel                         2.9.1
backcall                      0.2.0
bcrypt                        3.2.0
beautifulsoup4                4.11.1
binaryornot                   0.4.4
black                         21.12b0
bleach                        4.1.0
Bottleneck                    1.3.5
Brotli                        1.0.9
brotlipy                      0.7.0
bycycle                       1.0.0
certifi                       2022.6.15
cffi                          1.15.0
chardet                       4.0.0
charset-normalizer            2.0.10
chart-studio                  1.1.0
click                         8.0.4
cloudpickle                   2.0.0
colorama                      0.4.4
colour                        0.1.5
cookiecutter                  1.7.3
cryptography                  37.0.1
cycler                        0.11.0
dash                          2.0.0
dash-auth                     1.4.1
dash-bootstrap-components     0.13.1
dash-core-components          2.0.0
dash-html-components          2.0.0
dash-table                    5.0.0
deap                          1.3.1
debugpy                       1.5.1
decorator                     5.1.1
defusedxml                    0.7.1
diff-match-patch              20200713
docutils                      0.17.1
dtreeviz                      1.3.2
entrypoints                   0.4
executing                     0.8.3
explainerdashboard            0.3.7
fastjsonschema                2.15.1
flake8                        3.9.2
Flask                         2.0.2
Flask-Compress                1.10.1
Flask-SeaSurf                 0.3.1
flask-simplelogin             0.1.1
Flask-WTF                     0.15.1
fonttools                     4.29.0
future                        0.18.2
graphviz                      0.20.1
hypothetical                  0.3.0
idna                          3.3
imagesize                     1.3.0
importlib-metadata            4.11.3
inflection                    0.5.1
iniconfig                     1.1.1
intervaltree                  3.1.0
ipykernel                     6.9.1
ipython                       8.3.0
ipython-genutils              0.2.0
ipywidgets                    7.6.5
isort                         5.9.3
itsdangerous                  2.0.1
jedi                          0.18.1
Jinja2                        3.0.3
jinja2-time                   0.2.0
joblib                        1.1.0
jsonschema                    4.4.0
jupyter-client                7.1.2
jupyter-core                  4.10.0
jupyter-dash                  0.4.0
jupyterlab-pygments           0.1.2
jupyterlab-widgets            1.0.2
keyring                       23.4.0
kiwisolver                    1.3.2
lazy-object-proxy             1.6.0
lightgbm                      3.3.2
littleutils                   0.2.2
llvmlite                      0.38.0
lspopt                        1.1.1
MarkupSafe                    2.0.1
matplotlib                    3.5.1
matplotlib-inline             0.1.3
mccabe                        0.6.1
mistune                       0.8.4
mkl-fft                       1.3.1
mkl-random                    1.2.2
mkl-service                   2.4.0
mne                           0.24.1
mypy-extensions               0.4.3
nbclient                      0.5.13
nbconvert                     6.4.4
nbformat                      5.3.0
nest-asyncio                  1.5.5
neurodsp                      2.1.0
notebook                      6.4.7
numba                         0.55.0
numexpr                       2.8.3
numpy                         1.21.6
numpy-indexed                 0.3.5
numpydoc                      1.2
orjson                        3.6.6
outdated                      0.2.1
oyaml                         1.0
packaging                     21.3
pandas                        1.4.3
pandocfilters                 1.5.0
paramiko                      2.8.1
parso                         0.8.3
pathspec                      0.9.0
patsy                         0.5.2
pexpect                       4.8.0
philistine                    0.1
pickleshare                   0.7.5
Pillow                        9.0.0
pip                           21.2.4
platformdirs                  2.4.1
plotly                        5.5.0
pluggy                        1.0.0
poyo                          0.5.0
ppscore                       1.2.0
prometheus-client             0.12.0
prompt-toolkit                3.0.24
psutil                        5.8.0
ptyprocess                    0.7.0
pure-eval                     0.2.2
py                            1.11.0
py4j                          0.10.9.5
pycodestyle                   2.7.0
pycparser                     2.21
pydocstyle                    6.1.1
pyflakes                      2.3.1
Pygments                      2.11.2
pylint                        2.9.6
pyls-spyder                   0.4.0
PyNaCl                        1.4.0
pyOpenSSL                     22.0.0
pyparsing                     3.0.7
pyriemann                     0.2.7
pyrsistent                    0.18.1
PySocks                       1.7.1
pyspark                       3.3.0
pytest                        6.2.5
python-dateutil               2.8.2
python-lsp-black              1.0.0
python-lsp-jsonrpc            1.0.0
python-lsp-server             1.2.4
python-slugify                5.0.2
PythonMeta                    1.26
pytz                          2021.3
pywin32                       303
pywin32-ctypes                0.2.0
pywinpty                      1.1.6
PyYAML                        6.0
pyzmq                         22.3.0
QDarkStyle                    3.0.2
qstylizer                     0.1.10
QtAwesome                     1.0.3
qtconsole                     5.3.0
QtPy                          2.0.1
regex                         2022.3.15
requests                      2.27.1
retrying                      1.3.3
rope                          0.22.0
Rtree                         0.9.7
scikit-learn                  0.24.2
scipy                         1.7.3
seaborn                       0.11.2
Send2Trash                    1.8.0
setuptools                    58.0.4
shap                          0.40.0
sip                           4.19.13
six                           1.16.0
sleepecg                      0.5.2
slicer                        0.0.7
snowballstemmer               2.2.0
sortedcontainers              2.4.0
soupsieve                     2.3.1
Sphinx                        4.4.0
sphinxcontrib-applehelp       1.0.2
sphinxcontrib-devhelp         1.0.2
sphinxcontrib-htmlhelp        2.0.0
sphinxcontrib-jsmath          1.0.1
sphinxcontrib-qthelp          1.0.3
sphinxcontrib-serializinghtml 1.1.5
spyder                        5.1.5
spyder-kernels                2.1.3
stack-data                    0.2.0
statsmodels                   0.13.2
stochastic                    0.6.0
stopit                        1.1.2
tenacity                      8.0.1
tensorpac                     0.6.5
terminado                     0.12.1
testpath                      0.5.0
text-unidecode                1.3
textdistance                  4.2.1
threadpoolctl                 3.0.0
three-merge                   0.1.1
tinycss                       0.4
toml                          0.10.2
tomli                         1.2.3
torch                         1.10.2
tornado                       6.1
TPOT                          0.11.7
tqdm                          4.62.3
traitlets                     5.1.1
typed-ast                     1.4.3
typing_extensions             4.1.1
ua-parser                     0.10.0
ujson                         5.1.0
Unidecode                     1.2.0
update-checker                0.18.0
urllib3                       1.26.9
waitress                      2.0.0
watchdog                      2.1.6
wcwidth                       0.2.5
webencodings                  0.5.1
Werkzeug                      2.0.2
wheel                         0.37.1
widgetsnbextension            3.5.2
win-inet-pton                 1.1.0
wincertstore                  0.2
wrapt                         1.12.1
WTForms                       3.0.1
xgboost                       1.6.1
yapf                          0.31.0
yasa                          0.6.2
zipp                          3.8.0

**Comment by raphaelvallat** [comment, marked as answer]
Opened an issue here: https://github.com/raphaelvallat/yasa/issues/107 Let's move the conversation there
